list.c

```diff
@@ -184,6 +184,15 @@ void vListInsert( List_t * const pxList,
         *   4) Using a queue or semaphore before it has been initialised or
         *      before the scheduler has been started (are interrupts firing
         *      before vTaskStartScheduler() has been called?).
+        *     - This includes initializing binary semaphores before taking them. If
+        *       you create one with `xSemaphoreCreateBinary()` or
+        *       `xSemaphoreCreateBinaryStatic()`, you must call `xSemaphoreGive()`
+        *       before calling `xSemaphoreTake(). See:
+        *       https://freertos.org/xSemaphoreCreateBinaryStatic.html:
+        *       > The semaphore is created in the 'empty' state, meaning the
+        *       > semaphore must first be given using the xSemaphoreGive() API
+        *       > function before it can subsequently be taken (obtained) using the
+        *       > xSemaphoreTake() function.
```

Comment:
I don't want to expand this comment so much. Maybe point 4 could be changed to...

```
        *   4) Using a queue or semaphore before it has been initialised or
        *      before the scheduler has been started (are interrupts firing
        *      before vTaskStartScheduler() has been called?). See
        *      respective object documentation for more.
```

Reply:
I can see the only part you recommend adding then is this:

> See respective object documentation for more.

That doesn't add anything at all in my opinion. I've already been reading the respective object documentation for years, and actively reading it recently for months. 

---

## The scenario is this:

I'm in a 2 million line codebase. 
I did a 1.2 million line merge.

I have **no idea** why the code no longer runs, but I keep getting stuck in this for loop now and don't know why. What "respective object documentation" should I read!? We are using half of the FreeRTOS objects in the kernel and I don't know why my code won't run after a 1.2 million line merge!

By adding my two comments here, when debugging, I'd see this:

> IF YOU FIND YOUR CODE STUCK HERE, SEE THE NOTE JUST ABOVE.


...and then I'd read above. And when I read above, I'd then see this:

>         *     - This includes initializing binary semaphores before taking them. If
>         *       you create one with `xSemaphoreCreateBinary()` or
>         *       `xSemaphoreCreateBinaryStatic()`, you must call `xSemaphoreGive()`
>         *       before calling `xSemaphoreTake(). See:
>         *       https://freertos.org/xSemaphoreCreateBinaryStatic.html:
>         *       > The semaphore is created in the 'empty' state, meaning the
>         *       > semaphore must first be given using the xSemaphoreGive() API
>         *       > function before it can subsequently be taken (obtained) using the
>         *       > xSemaphoreTake() function.

and then I'd go search my merge commit for changes to semaphores, see we converted a mutex to a binary semaphore, and add the give call after creating it. Boom done. Would have saved me **2 weeks** of non-stop debugging.